### ISSUE TITLE
Extract data on monthly basis

### ISSUE THREAD

**Issue by frinyvonnick** [zenika-open-source/insights-website#9]
We should extract and store data from github on a monthly basis so we can do comparisons over the time. I made a tool with @jlandure called [github-insights](https://github.com/Zenika/github-insights) that use github API to achieve it. 

*Ideas:* Use a cloud function to extract and store data. @hgwood could give some advices about it.

**Comment by hgwood** [comment]
Using Google Cloud, the tool could be deployed as a Cloud Function, called periodically using Task Scheduler, and the resulting data could be stored in Cloud Storage as JSON files. Question: if the web site is static, how is it updated with the new monthly data? Should it be rebuilt and redeployed?

**Comment by mfrachet** [comment]
It depends on why you're asking it 😛 

JAMStack is good a building static sites, but it also allows to make some client side fetching like any JS app.

However if every pages of the app are "dynamics" (in the sense that there can be N version of one page using `/resource/:id`) JAMStack is not a good fit because of SEO & First Meaningful Paint.

I'm not sure it's a big deal for the FMP in this app since we will probably cache the value we get for some times?

**Comment by kefranabg** [comment]
If the webapp dynamic data does not require frequent updates, I think JAMStack architecture can fit our needs. We'll just need to rebuild & deploy when new data is added/updated :)

**Comment by frinyvonnick** [comment]
With @bpetetot we hesitate about updating data on a monthly or weekly basis. I think monthly is enough but it's open to discussion 😄

**Comment by frinyvonnick** [comment]
@mfrachet we want only one page but update data frequently and make comparison with previous data so we will re-build the website each time we update data. Is that clearer ?

**Comment by christopherkade** [comment]
Just to be sure I'm understanding the execution flow, the data would be extracted monthly (or weekly) by our Cloud Function and stored using Cloud Storage, which would then trigger a new build in order to retrieve this new data and generate the static website. Am I correct?

**Comment by frinyvonnick** [comment]
That's correct @christopherkade 👍 !

**Comment by christopherkade** [comment]
I got a chance to test https://github.com/Zenika/github-insights this morning to see what format our open-source data would take, namely one JSON file per member of the organisation.

Which made me wonder, is it necessary to keep all these files, or should we run them through an adapter to only keep what’s relevant? (Number of repos contributed too, total number of contributions etc.)

**Comment by frinyvonnick** [comment]
An adapter is quite a good idea. Do you have something specific in mind ?

**Comment by christopherkade** [comment]
Not at the moment, although I'm working on something quite similar in my current project as an intern here. Where we extract data from a CMS (Prismic) on each build and adapt it to our needs by simply creating and filling up a JSON file. 

We could do something similar, although I'm not sure of the merit of having it as a Cloud Function if it must be executed on each build. Why not have the script associated to the project's repository and executed on each build?

**Comment by frinyvonnick** [comment]
Sounds like a good idea ! Do you want to do it in a PR @christopherkade ?

**Comment by christopherkade** [comment]
I could give it a try @frinyvonnick 😄 

I noticed that [Github Insights](https://github.com/Zenika/github-insights) has a `stats` command that already adapts the data (but only outputs it in the console). We could run the script, then execute the stats command and put the output into a file. That would imply tweaking the existing work and perhaps deploying it as a package. 

What do you think?

**Comment by frinyvonnick** [comment]
Nice idea could you make a proof of concept ?

**Comment by christopherkade** [comment]
Of course, I did one on the following PR: https://github.com/Zenika/github-insights/pull/5

**Comment by kefranabg** [comment]
On [GCP](https://console.cloud.google.com), we could use the [cloud scheduler](https://cloud.google.com/scheduler/) tool that periodically fires a cloud function. Already used this combination for a client project, works perfectly 👌  

Once the data retrieved, open a pull request with [github pull request api](https://developer.github.com/v3/pulls/#create-a-pull-request).
